test/python/circuit/test_library.py

```diff
@@ -50,3 +58,449 @@ def test_inner_product(self):
+                value += prob
+                # all other states should have zero probability due to ancilla qubits
+                if i > 2**num_req_qubits:
+                    break
+
+        # map value to original range
+        mapped_value = asian_barrier_spread.value_to_estimation(
+            value) / (2**num_uncertainty_qubits - 1) * (high_ - low_)
+        expected = 0.83188
+        self.assertAlmostEqual(mapped_value, expected, places=5)
+
+
+@ddt
+class TestWeightedAdder(QiskitTestCase):
+    """Test the weighted adder circuit."""
+
+    def assertSummationIsCorrect(self, adder):
+        """Assert that ``adder`` correctly implements the summation w.r.t. it's set weights."""
+
+        circuit = QuantumCircuit(adder.num_qubits)
+        circuit.h(list(range(adder.num_state_qubits)))
+        circuit.append(adder.to_instruction(), list(range(adder.num_qubits)))
+
+        backend = BasicAer.get_backend('statevector_simulator')
+        statevector = execute(circuit, backend).result().get_statevector()
+
+        probabilities = defaultdict(float)
+        for i, statevector_amplitude in enumerate(statevector):
+            i = bin(i)[2:].zfill(circuit.num_qubits)[adder.num_ancilla_qubits:]
+            probabilities[i] += np.real(np.abs(statevector_amplitude) ** 2)
+
+        expectations = defaultdict(float)
+        for x in range(2**adder.num_state_qubits):
+            bits = np.array(list(bin(x)[2:].zfill(adder.num_state_qubits)), dtype=int)
+            summation = bits.dot(adder.weights[::-1])
+
+            entry = bin(summation)[2:].zfill(adder.num_sum_qubits) \
+                + bin(x)[2:].zfill(adder.num_state_qubits)
+            expectations[entry] = 1 / 2 ** adder.num_state_qubits
+
+        for state, probability in probabilities.items():
+            self.assertAlmostEqual(probability, expectations[state])
+
+    @data(
+        [0],
+        [1, 2, 1],
+        [4],
+    )
+    def test_summation(self, weights):
+        """Test the weighted adder on some examples."""
+        adder = WeightedAdder(len(weights), weights)
+        self.assertSummationIsCorrect(adder)
+
+    def test_mutability(self):
+        """Test the mutability of the weighted adder."""
+        adder = WeightedAdder()
+
+        with self.subTest(msg='missing number of state qubits'):
+            with self.assertRaises(AttributeError):
+                print(adder.draw())
+
+        with self.subTest(msg='default weights'):
+            adder.num_state_qubits = 3
+            default_weights = 3 * [1]
+            self.assertListEqual(adder.weights, default_weights)
+
+        with self.subTest(msg='specify weights'):
+            adder.weights = [3, 2, 1]
+            self.assertSummationIsCorrect(adder)
+
+        with self.subTest(msg='mismatching number of state qubits and weights'):
+            with self.assertRaises(ValueError):
+                adder.weights = [0, 1, 2, 3]
+                print(adder.draw())
```

Comment:
does the testsuite need to draw the circuit?

Reply:
see the answer to the same question above